providers/implementations/ciphers/cipher_aes_gcm_siv.c

```diff
@@ -0,0 +1,326 @@
+    if (ret == NULL) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+    /* NULL-out these things we create later */
+    ret->aad = NULL;
+    ret->ecb_ctx = NULL;
+
+    if (in->aad == NULL) {
+        if ((ret->aad = OPENSSL_memdup(in->aad, UP16(ret->aad_len))) == NULL) {
+            ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+            goto err;
+        }
+    }
+
+    if (!in->hw->dup_ctx(ret, in))
+        goto err;
+
+    return ret;
+ err:
+    if (ret != NULL) {
+        OPENSSL_clear_free(ret->aad, ret->aad_len);
+        OPENSSL_free(ret);
+    }
+    return NULL;
+}
+
+static int ossl_aes_gcm_siv_init(void *vctx, const unsigned char *key, size_t keylen,
+                                 const unsigned char *iv, size_t ivlen,
+                                 const OSSL_PARAM params[], int enc)
+{
+    PROV_AES_GCM_SIV_CTX *ctx = (PROV_AES_GCM_SIV_CTX *)vctx;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    ctx->enc = enc;
+
+    if (key != NULL) {
+        if (keylen != ctx->key_len) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH);
+            return 0;
+        }
+        memcpy(ctx->key_gen_key, key, ctx->key_len);
+    }
+    if (iv != NULL) {
+        if (ivlen != sizeof(ctx->nonce)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_IV_LENGTH);
+            return 0;
+        }
+        memcpy(ctx->nonce, iv, sizeof(ctx->nonce));
+    }
+
+    if (!ctx->hw->initkey(ctx))
+        return 0;
+
+    return ossl_aes_gcm_siv_set_ctx_params(ctx, params);
+}
+
+static int ossl_aes_gcm_siv_einit(void *vctx, const unsigned char *key, size_t keylen,
+                                  const unsigned char *iv, size_t ivlen,
+                                  const OSSL_PARAM params[])
+{
+    return ossl_aes_gcm_siv_init(vctx, key, keylen, iv, ivlen, params, 1);
+}
+
+static int ossl_aes_gcm_siv_dinit(void *vctx, const unsigned char *key, size_t keylen,
+                                  const unsigned char *iv, size_t ivlen,
+                                  const OSSL_PARAM params[])
+{
+    return ossl_aes_gcm_siv_init(vctx, key, keylen, iv, ivlen, params, 0);
+}
+
+#define ossl_aes_gcm_siv_stream_update ossl_aes_gcm_siv_cipher
+static int ossl_aes_gcm_siv_cipher(void *vctx, unsigned char *out, size_t *outl,
+                                   size_t outsize, const unsigned char *in, size_t inl)
+{
+    PROV_AES_GCM_SIV_CTX *ctx = (PROV_AES_GCM_SIV_CTX *)vctx;
+    int error = 0;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    /* The RFC has a test case for this, but we don't try to do anything */
+    if (inl == 0) {
+        if (outl != NULL)
+            *outl = 0;
+        return 1;
+    }
+
+    if (outsize < inl) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
+        return 0;
+    }
+
+    error |= !ctx->hw->cipher(ctx, out, in, inl);
+
+    if (outl != NULL)
+        *outl = inl;
```

Comment:
I am not sure I like this construction. I'd prefer not modifying the *outl in case of error.

Reply:
Despite the `if`, this is attempting to be constant-time processing, but not sure how important that is in the grand scope of things.

Reply:
I updated it because it is an error...